acvm_backend_barretenberg/src/composer.rs

```diff
@@ -167,8 +198,235 @@ impl StandardComposer {
+        let proof_ptr_ptr: usize = 0;
+
+        let proof_size = self
+            .call_multiple(
+                "acir_proofs_new_proof",
+                vec![
+                    &pippenger_ptr,
+                    &g2_ptr,
+                    &pk_ptr,
+                    &cs_ptr,
+                    &witness_ptr,
+                    &Value::I32(0),
+                ],
+            )
+            .value();
+        let proof_size: usize = proof_size.unwrap_i32() as usize;
+
+        // We then need to read the pointer at `proof_ptr_ptr` to get the proof's location
+        // and then slice memory again at `proof_ptr` to get the proof data.
+        let proof_ptr = self.slice_memory(proof_ptr_ptr, POINTER_BYTES);
+        let proof_ptr: usize =
+            u32::from_le_bytes(proof_ptr[0..POINTER_BYTES].try_into().unwrap()) as usize;
+
+        let result = self.slice_memory(proof_ptr, proof_size);
+
+        proof::remove_public_inputs(constraint_system.public_inputs_size(), &result)
+    }
+
+    fn verify_with_vk(
+        &self,
+        constraint_system: &ConstraintSystem,
+        // XXX: Important: This assumes that the proof does not have the public inputs pre-pended to it
+        // This is not the case, if you take the proof directly from Barretenberg
+        proof: &[u8],
+        public_inputs: Assignments,
+        verification_key: &[u8],
+    ) -> bool {
+        use wasmer::Value;
+
+        let circuit_size = self.get_circuit_size(constraint_system);
+        let CRS { g2_data, .. } = CRS::new(circuit_size as usize);
+
+        // Prepend the public inputs to the proof.
+        // This is how Barretenberg expects it to be.
+        // This is non-standard however, so this Rust wrapper will strip the public inputs
+        // from proofs created by Barretenberg. Then in Verify we prepend them again.
+        //
+        let proof = proof::prepend_public_inputs(proof.to_vec(), public_inputs);
+        let cs_buf = constraint_system.to_bytes();
+
+        let cs_ptr = self.allocate(&cs_buf);
+        let proof_ptr = self.allocate(&proof);
+        let g2_ptr = self.allocate(&g2_data);
+        let vk_ptr = self.allocate(verification_key);
+
+        let verified = self
+            .call_multiple(
+                "acir_proofs_verify_proof",
+                vec![
+                    &g2_ptr,
+                    &vk_ptr,
+                    &cs_ptr,
+                    &proof_ptr,
+                    &Value::I32(proof.len() as i32),
+                ],
+            )
+            .value();
+
+        self.free(proof_ptr);
+
+        match verified.unwrap_i32() {
+            0 => false,
+            1 => true,
+            _ => panic!("Expected a 1 or a zero for the verification result"),
+        }
+    }
+}
+
 fn pow2ceil(v: u32) -> u32 {
-    v.next_power_of_two()
+    #[cfg(feature = "native")]
+    {
+        v.next_power_of_two()
+    }
```

Comment:
Why use this helper at all? Seems like the `not(native)` impl would work for both

Reply:
Sure, this is all private so swapping out the implementation later is nbd.

Let's avoid doing so many checks though as we know the value for `v` which will begin to overflow

```rust
if v > (u32::MAX >> 1) {
    panic!("pow2ceil overflow");
}

let mut p = 1;
while p < v {
    p <<= 1;
}
p
```

Reply:
sounds good to me 🎉